### ISSUE TITLE
Running Network with Conv2D Layers on GPU produces unexpected results.

### ISSUE THREAD

**Issue by LunaticWyrm467** [huggingface/candle#575]
### The Problem
I have a neural network with the following setup:
```
In:  Conv2D -> ReLU
     MaxPool2D
     Conv2D -> ReLU
     MaxPool2D
Out: Linear -> SiLU
```

When I use the CPU, my neural network works fine, and displays different outputs depending on the inputs given to it.
When I use the GPU on the other hand, my network always outputs the same values, no matter the given inputs.

### Environment
- Windows 11
- 11th Gen Intel Core i5-11260H
- 8.0 GB DDR4 RAM 
- NVidia GeForce RTX 3050 Laptop GPU
- CUDA 11.8 Installed

### MRP
Here is a small MRP with a .toml project configuration file and a single main.rs program file.
At line 126, you're able to set `use_cpu` to true to run the control test.
The program itself runs an infinite loop at which random tensors are feed into the network, and the outputs are printed to the console.
[mrp.zip](https://github.com/huggingface/candle/files/12424470/mrp.zip)

**Comment by LaurentMazare** [comment]
Thanks a lot for reporting this, turns out this was caused by a bug in the conv2d kernel that only triggered on non-square inputs (roughly assuming NCWH layout rather than NCHW), this should be fixed by #579 and a test matching against PyTorch has been added.
Please give a try to the latest version and let us know how it gets.

**Comment by LunaticWyrm467** [comment]
Oh wow, glad to know there's already a solution in place! Will be sure to test the new version when I have time later today.

**Comment by LaurentMazare** [comment]
Yes please, and feel free to open new issues if you spot anything else being weird.
